lib/js/jsffi.nim

```diff
@@ -0,0 +1,391 @@
+  ## importing globals or things like `self` or `this` from JavaScript.
+  ## It takes the name of the variable and its type.
+  ##
+  ## Example:
+  ##
+  ## .. code-block:: nim
+  ##
+  ##  # We want to import the console object as an JsObject
+  ##  magicVar(console, JsObject)
+  ##  console.log("Hello JavaScript!")
+  ##
+  var `name` {. importc, nodecl, inject .}: `typ`
+
+# New
+proc newJsObject*: JsObject = {. emit: [result, " = {};"] .}
+  ## Creates a new empty JsObject
+proc newJsAssoc*[K, V]: JsAssoc[K, V] = {. emit: [result, " = {};"] .}
+  ## Creates a new empty JsAssoc with key type `K` and value type `V`.
+
+# Checks
+proc hasOwnProperty*(x: JsObject, prop: string): bool
+  {. importcpp: "#.hasOwnProperty(#)" .}
+  ## Checks, whether `x` has a property of name `prop`.
+
+proc jsTypeOf*(x: JsObject): cstring {. importcpp: "typeof(#)" .}
+  ## Returns the name of the JsObject's JavaScript type as a cstring.
+
+# Conversion to and from JsObject
+proc to*(x: JsObject, T: typedesc): T {. importcpp: "(#)" .}
+  ## Converts a JsObject `x` to type `T`.
+proc toJs*[T](val: T): JsObject {. importcpp: "(#)" .}
+  ## Converts a value of any type to type JsObject
+
+# Field accessors for JsObject and JsAssoc types
+proc compareImpl(x, y: JsRoot): bool {. importcpp: "(# == #)" .}
+
+proc setImpl[K, V](obj: JsRoot, field: K, val: V)
+  {.importcpp: "#[#] = #" .}
+
+proc getAsImpl[K, V](obj: JsRoot, field: K): V
+  {. importcpp: "#[#]" .}
+
+proc callFieldAsImpl[K, V](obj: JsRoot, field: K, args: varargs[JsObject]): V =
+  when V isnot void:
+    {. emit: [result, "=", obj, "[", field, "].apply(", obj, ", ", args, ");"] .}
+  else:
+    {. emit: [obj, "[", field, "].apply(", obj, ", ", args, ");"] .}
+
+proc `[]`*(obj: JsObject, field: cstring): JsObject =
+  ## Return the value of a property of name `field` from a JsObject `obj`.
+  getAsImpl[cstring, JsObject](obj, field)
+
+proc `[]=`*[T](obj: JsObject, field: cstring, val: T) =
+  ## Set the value of a property of name `field` in a JsObject `obj` to `v`.
+  setImpl[cstring, JsObject](obj, field, val.toJs)
+
+proc `[]`*[K, V](obj: JsAssoc[K, V], field: K): V =
+  ## Return the value of a property of name `field` from a JsAssoc `obj`.
+  when K is string:
+    getAsImpl[cstring, V](obj, field)
+  else:
+    getAsImpl[K, V](obj, field)
+
+proc `[]=`*[K, V](obj: JsAssoc[K, V], field: K, val: V) =
+  ## Set the value of a property of name `field` in a JsAssoc `obj` to `v`.
+  when K is string:
+    setImpl[cstring, V](obj, field, val)
+  else:
+    setImpl[K, V](obj, field, val)
+
+proc `==`*(x, y: JsRoot): bool =
+  ## Compare two JsObjects or JsAssocs. Be careful though, as this is comparison
+  ## like in JavaScript, so if your JsObjects are in fact JavaScript Objects,
+  ## and not strings or numbers, this is a *comparison of references*.
+  compareImpl(x, y)
+
+{. experimental .}
+proc `.`*(obj: JsObject, field: cstring): JsObject =
```

Comment:
This should better use dot-notation in the resulting codegen and be a template ;)